### PR TITLE
Mmf/async plugins

### PR DESCRIPTION
**Pull request by mfrazier63**
Make the monitor and detector plugins asynchronous using CompletableFuture.

Added a unit test for ICMP. Not SNMP for now as it uses a static method on a util class!

### CODE REVIEW

File: ignite-worker/src/main/java/org/opennms/poc/ignite/worker/ignite/service/WorkflowService.java [hunk 1]
```diff
@@ -55,6 +64,11 @@ public void execute() {
         log.info("########### registered detector count {}", OsgiServiceHolder.getRegisteredDetectorCount());
         Optional<ServiceDetector> detector = OsgiServiceHolder.getDetector(workflow.getType());
         log.info("Detector is {}", detector.isPresent() ? detector.toString():"NOT FOUND");
+        Optional<ServiceMonitor> serviceMonitor = OsgiServiceHolder.getMonitor("blah");
+
+        for (int i=0;i< 5;i++) {
+            serviceMonitor.get().poll(null ,null).whenComplete((serviceMonitorResponse, exception ) -> log.info("got a response") );
```
Comment (j-white) [inline ignite-worker/src/main/java/org/opennms/poc/ignite/worker/ignite/service/WorkflowService.java:70]: @artnaseef this good for now, or should do we something different here?

Comment (artnaseef) [reply ignite-worker/src/main/java/org/opennms/poc/ignite/worker/ignite/service/WorkflowService.java:70]: @j-white ah, I think this is touching on a topic I neglected to raise.  The `load-em-up` REST endpoint uses different Ignite services than the Workflow driven by Twin updates.  Do we want to merge them, or perhaps drop the REST endpoint and this class altogether?  Responsibility for this mess lies with me.

Comment (mfrazier63) [reply ignite-worker/src/main/java/org/opennms/poc/ignite/worker/ignite/service/WorkflowService.java:70]: This class is deprecated now anyway.

File: plugins/ICMP/src/main/java/org/opennms/poc/icmp/IcmpMonitor.java [hunk 2]
```diff
@@ -1,59 +1,52 @@
...
-            LOG.debug("failed to ping {}", host, e);
-            return PollStatus.unavailable(e.getMessage());
-        }
-        
-        if (rtt != null) {
-            return PollStatus.available(rtt.doubleValue());
-        } else {
-            // TODO add a reason code for unavailability
-            return PollStatus.unavailable(null);
+            pinger.ping(null, 1, 2, 3, new PingResponseCallback() {
```
[the remark this answers is not in the document]
Comment (mfrazier63) [reply plugins/ICMP/src/main/java/org/opennms/poc/icmp/IcmpMonitor.java:29]: Already fixed this while double checking tests.

File: plugins/ICMP/src/main/java/org/opennms/poc/icmp/IcmpMonitor.java [hunk 3]
```diff
@@ -1,59 +1,52 @@
...
+                @Override
+                public void handleResponse(InetAddress inetAddress, EchoPacket response) {
+                    double responseTimeMicros = Math.round(response.elapsedTime(TimeUnit.MICROSECONDS));
+                    //TODO: set the responseTimeMicros into the response properties?
```
[the remark this answers is not in the document]
Comment (mfrazier63) [reply plugins/ICMP/src/main/java/org/opennms/poc/icmp/IcmpMonitor.java:33]: - [ ] set the properties

File: plugins/ICMP/src/main/java/org/opennms/poc/icmp/IcmpMonitor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mfrazier63) [reply plugins/ICMP/src/main/java/org/opennms/poc/icmp/IcmpMonitor.java:29]: - [ ] fix pinger params

File: plugins/ICMP/src/main/java/org/opennms/poc/icmp/IcmpMonitor.java [hunk 4]
```diff
@@ -1,59 +1,52 @@
...
-        pingerFactory = Suppliers.memoize(() -> pingerFactoryDelegate);
-    }
-    
     @Override
-    public PollStatus poll(MonitoredService svc, Map<String, Object> parameters) {
-        Number rtt = null;
-        InetAddress host = svc.getAddress();
+    public CompletableFuture<ServiceMonitorResponse> poll(MonitoredService svc, Map<String, Object> parameters) {
 
+        Pinger pinger = pingerFactory.getInstance();
```
[the remark this answers is not in the document]
Comment (mfrazier63) [reply plugins/ICMP/src/main/java/org/opennms/poc/icmp/IcmpMonitor.java:26]: - [ ] change pingerFactory call

**Review by artnaseef — APPROVED** [review]

**Pull request merged**